### PR TITLE
Unstuck feature

### PR DESCRIPTION
**Pull request by alecsci**
# Feature: Players can now login and unstuck their characters

## Flow
- Page is visible below the Characters page
image
- Click on the hearthstone button and check the in-game character 

## Technical implementation
- Relied on how character page load works (used the same styling, load hook, css class to render characters) + an extra column for the hearthstone button 
- Created separate REST POST endpoint to handle the unstuck operation
- Unstuck does some validation (making sure is the same user which is logged in, if the character name exists, if the account exists otherwise throw 4xx)
- Unstuck updates the hearthstone spell cd in DB
- Unstuck disables the hearthstone button and display a tooltip on hover
- After unstuck page reloads to reflect the button state
- Countdown timer shown until CD is lifted

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -87,12 +89,16 @@ services:
...
+      - ac-network  # Add this line to connect to the ac-network
 networks:
   local-shared-net:
     name: local-shared-net
     driver: bridge
   local-private-net:
     driver: bridge
+  ac-network:
+    name: azerothcore-wotlk_ac-network
+    external: true  # Add this line to use the ac-network defined in the other docker-compose file
```
Comment (Helias) [inline docker-compose.yml:101]: I would not change this, the name of the network is custom and it is not necessary if you are not using docker with azerothcore, moreover there is already the documentation about using acore-cms+docker with azerothcore+docker

File: srv/wordpress/wp-content/plugins/acore-wp-plugin/src/Manager/Soap/UnstuckService.php [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+use ACore\Manager\Soap\AcoreSoapTrait;
+
+class UnstuckService
+{
+
+    use AcoreSoapTrait;
+
+    public function unstuckByName($charName)
+    {
+        return $this->executeCommand(".unstuck $charName");
```
Comment (Helias) [inline srv/wordpress/wp-content/plugins/acore-wp-plugin/src/Manager/Soap/UnstuckService.php:14]: nice

File: srv/wordpress/wp-content/plugins/acore-wp-plugin/src/Components/UnstuckMenu/UnstuckController.php [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+        $query = "
+            INSERT INTO `character_spell_cooldown` (`guid`, `spell`, `time`)
+            SELECT c.`guid`, 8690, $newTime
+            FROM `characters` c
+            WHERE c.`name` = '$charName'
+            AND c.`deleteDate` IS NULL
+            AND c.`account` = $accId
+            ON DUPLICATE KEY UPDATE `time` = `character_spell_cooldown`.`time`
+        ";
```
Comment (Helias) [inline srv/wordpress/wp-content/plugins/acore-wp-plugin/src/Components/UnstuckMenu/UnstuckController.php:79]: good job!

**Review by Helias — APPROVED** [review]

File: srv/wordpress/wp-content/plugins/acore-wp-plugin/src/Components/UnstuckMenu/UnstuckView.php [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            <div class="col-sm-4">
+                <div class="card">
+                    <div class="card-body">
+                        <h3>Unstuck</h3>
+                        <p>Unstuck your characters and teleport them to the hearthstone location</p>
+                        <hr>
+                        <ul id="acore-characters-unstuck" class="list-unstyled">
+                            <?php foreach ($chars as $char) {
+                                $currentTime = time();
+                                $isDisabled = ($char["time"] > $currentTime);
```
Comment (alecsci) [inline srv/wordpress/wp-content/plugins/acore-wp-plugin/src/Components/UnstuckMenu/UnstuckView.php:71]: Encountered edge-case: the buttons are never disabled when user used a day before the HS, but the DB doesn't get immediately updated.

1. use hs, gets cd, state is not yet persisted in DB
2. logout, state is persisted in DB
3. wait 1h, CD should be down
4. login, DB still has the CD even if it's in the past, but the client is allowing the player to use hs
5. logout (the CD is removed from DB)

File: srv/wordpress/wp-content/plugins/acore-wp-plugin/src/Components/UnstuckMenu/UnstuckController.php [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        $newTime = $newTime = time() + (15 * 60); // 15 minutes;
+
+        $query = "
+            INSERT INTO `character_spell_cooldown` (`guid`, `spell`, `time`)
+            SELECT c.`guid`, 8690, $newTime
+            FROM `characters` c
+            WHERE c.`name` = '$charName'
+            AND c.`deleteDate` IS NULL
+            AND c.`account` = $accId
+            ON DUPLICATE KEY UPDATE `time` = `character_spell_cooldown`.`time`
```
Comment (alecsci) [inline srv/wordpress/wp-content/plugins/acore-wp-plugin/src/Components/UnstuckMenu/UnstuckController.php:78]: I should set the value for time from the $newTime, not with the existing column value.

File: srv/wordpress/wp-content/plugins/acore-wp-plugin/web/assets/unstuck/unstuck.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+            }
+        });
+    });
+});
+
+
+// Method to update the countdown of characters that used the unstuck option
+document.addEventListener('DOMContentLoaded', function() {
+    function updateCountdown() {
+        var now = Math.floor(Date.now() / 1000); // Get current time in seconds
```
Comment (Helias) [inline srv/wordpress/wp-content/plugins/acore-wp-plugin/web/assets/unstuck/unstuck.js:26]: never use var, only `const` or `let`

**Pull request merged**